### ISSUE TITLE
method_missing doesn't get arguments

### ISSUE THREAD

**Issue by jtara** [mustache/mustache#176]
Trying to implement some logic using `method_missing`. (I know, slap my hand!)

Using Ruby 2.1.2 and Mustache 0.99.5

To clear-up confusion, I am processing Ruby files as my "templates". (disabled HTML escape). I'm using it as a preprocessor.

Example:

``` ruby
        {{#if_product_is_miscore_c}}
          menu << { :name => :upload_ceas,   :icon => :'upload'}  # REMOVEME: for test
        {{/if_product_is_miscore_c}}
```

Part of my view:

``` ruby
class PrepareView < Mustache

  HTML_EXTENSIONS = Set.new [
    '.html',
    '.erb'
  ]

  def initialize(template_path_name, product_dir, data)
    self.template_file = template_path_name
    @product_dir = product_dir
    @data = data
  end

  def method_missing(sym, *arguments, &block)
    puts "method_missing sym = #{sym.inspect} arguments = #{arguments.inspect} $1 = #{$1}"
    if block_given?
      puts "method_missing block given #{block.inspect}"
    end
    match_data = /if_product_is_(.*)/.match sym.to_s
    if match_data
      puts "method_missing matched if_product_is_"
      product = match_data[1]
      puts "method_missing product = #{product.inspect} product.underscore = #{product.underscore.inspect}"
      if product == @data['build']['product'].underscore
        puts "method_missing matched product #{product.inspect} #{arguments.first.inspect}"
        return $1
      else
        super
      end
    else
      super
    end
  end

  def respond_to_missing?(sym, include_private=false)
    puts "respond_to_missing? sym = #{sym.inspect} include_private = #{include_private.inspect}"
    match_data = /if_product_is_(.*)/.match sym.to_s
    if match_data
      puts "respond_to_missing? matched if_product_is"
      product = match_data[1]
      puts "respond_to_missing? product = #{product.inspect} product.underscore = #{product.underscore.inspect} build.product = #{@data['build']['product'].inspect}"
      if product == @data['build']['product'].underscore
        puts "respond_to_missing? matched product #{product.inspect}"
        return true
      else
        puts "respond_to_missing? did not match product #{product.inspect}"
        super
      end
    else
      puts "respond_to_missing? did not match if_product_is_"
      super
    end
  end

  # Mustache normally HTML-escapes content, unless you use triple braces {{{ugly}}} or & {{&awkward}}
  # But most of the files we will process are not HTML.
  # So, bypass escape unless it is an HTML source file
  def escapeHTML(txt)
    if HTML_EXTENSIONS.include? File.extname(template_path)
      super.escapeHTML txt
    else
      txt
    end
  end
```

The view is used like this:

``` ruby
          puts "#{product_file_path} #{MUSTACHE_ARROW} #{target_file_path}"
          data = load_file_data product_file_path, framework_file_path
          view = PrepareView.new product_file_path, @product_dir, data
          File.open target_file_path, 'w' do |f|
            f.puts view.render data
          end
```

Some diagnostic output:

```
framework/app/helpers/form_helper.rb -> build/ios/miscore-c/development/simulator/7.1/app/helpers/form_helper.rb
respond_to_missing? sym = :has_key? include_private = false
respond_to_missing? did not match if_product_is_
respond_to_missing? sym = :if_product_is_miscore_c include_private = false
respond_to_missing? matched if_product_is
respond_to_missing? product = "miscore_c" product.underscore = "miscore_c" build.product = "miscore-c"
respond_to_missing? matched product "miscore_c"
respond_to_missing? sym = :if_product_is_miscore_c include_private = true
respond_to_missing? matched if_product_is
respond_to_missing? product = "miscore_c" product.underscore = "miscore_c" build.product = "miscore-c"
respond_to_missing? matched product "miscore_c"
method_missing sym = :if_product_is_miscore_c arguments = [] $1 =
method_missing matched if_product_is_
method_missing product = "miscore_c" product.underscore = "miscore_c"
method_missing matched product "miscore_c" nil
```

I'd expected to get the enclosed text in `arguments.first` (or `$1`). I'm getting empty arguments.

What am I missing here?

A normally-defined method does work as I would expect:

``` ruby
  # Render an optional partial from the product directory
  # This differs from regular Mustache partials:
  # - it is optional, if the file does not exist, it does not throw an error
  # - it prepends the product directory to the supplied path
  # - syntax by necessity differs
  #
  # Use like:
  #
  # ({#product_partial}}
  #   rhoconfig_product.txt
  #  {{/product_partial}}
  def product_partial(f_in)
    puts "product partial f_in = #{f_in.inspect}"
    f = "#{@product_dir}/#{f_in.strip}.mustache"
    File.exists?(f) ? File.read(f) : ''
  end
```

**Comment by hugopeixoto** [comment]
TLDR: implement #method_missing so that it returns a 1-arity proc. Mustache will call it with the enclosed text as its first argument, and it should work fine.

The problem is that mustache handles context method calls in two very different ways, depending on arity. When in a section tag, if the arity is 1, the method is immediately treated as a Proc (which receives the enclosed text). If the arity is not 1, the method is called and its return value is then checked for Proc/truth/list-iness.

When you use method missing, Mustache cannot detect the arity of the method, so it assumes that it is a zero arity method, and you end up in the second case.

So, as a workaround you could use:

    class MethodMissingView < Mustache
      def method_missing method, *args, &block
        if method == :things
          proc { |text| text.upcase }
        else
          super
        end
      end

      def respond_to_missing?(method, *)
        method == :things || super
      end
    
      def to_s; 'method_missing' end
    end
    
    MethodMissingView.render('{{#things}}current context is {{.}}{{/things}}')
    #=> "CURRENT CONTEXT IS method_missing"
